### PR TITLE
Rename bookmarks

### PR DESCRIPTION
**Pull request by jyothishnt**
## Type

- [ ] Bug fix
- [ ] New feature
- [x] Improvement to existing feature

## Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1119

## Description
The section we call 'Bookmarks' should be renamed 'Previously viewed' (we have used this phrase elsewhere, and it is an more accurate description of what is being displayed).

Also remove the 'Example links' section

## Deployment URL
http://rename-bookmarks.review.ensembl.org/

## Views affected
GB and EV

### Other effects
N/A

- [ ] I have checked any requirements for Google Analytics
- [ ] I have created any required tests
- [ ] The PR has as its final commit a WASM/JS update commit if any Rust files were updated.

## Possible complications
Nothing I am aware of. Though I have noticed bookmarks being closed when interacting with browser chrome. Will have to fix this in another ticket.

### CODE REVIEW

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.scss [hunk 1]
```diff
@@ -20,24 +24,28 @@
...
 }
 
+.previouslyViewedVersionedStableId {
+  margin-left: 8px;
+}
+
 .title {
   font-size: 14px;
-  margin: 5px 0 10px 10px;
+  margin: 5px 0 25px 0px;
```
Comment (imransl) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.scss:33]: We can remove `px` from `0px`

File: src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts [hunk 2]
```diff
@@ -144,22 +134,29 @@ export const updatePreviouslyViewedObjectsAndSave = (): ThunkAction<
...
+    object_id: activeEnsObject.object_id,
+    object_type: activeEnsObject.type,
+    label: activeEnsObject.label,
+    versioned_stable_id:
+      activeEnsObject.type === 'gene'
+        ? activeEnsObject.versioned_stable_id
+        : null
+  };
+
+  const updatedObject = [newObject, ...savedEntitiesWithoutCurrentEntity].slice(
```
Comment (imransl) [inline src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts:153]: Why is `Object` in the name of this variable when it actually is an array?

File: src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts [hunk 3]
```diff
@@ -144,22 +134,29 @@ export const updatePreviouslyViewedObjectsAndSave = (): ThunkAction<
...
 
-  const existingIndex = previouslyViewedObjects.findIndex(
-    (previouslyViewedObject) =>
-      previouslyViewedObject.object_id === activeEnsObject.object_id
+  const savedEntitiesWithoutCurrentEntity =
+    previouslyViewedObjects?.filter(
+      (entity) => entity.object_id !== activeEnsObject.object_id
+    ) || [];
+
+  const newObject = {
```
Comment (imransl) [inline src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts:142]: Is this `newObject` or `defaultObject`?

File: src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts [hunk 4]
```diff
@@ -143,175 +140,163 @@ export const clearSelectedSearchResult = createAction(
...
-  }
-
-  const newCommittedSpecies = [
-    ...committedSpecies,
-    {
-      ...pickBy(genomeInfo, (value, key) => {
-        return key !== 'example_objects';
-      }),
-      isEnabled: true
+export const ensureSpeciesIsCommitted =
```
Comment (imransl) [inline src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts:143]: How about changing this too: `ensureSpeciesIsCommitted` -> `areSpeciesCommitted`?

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.scss [hunk 5]
```diff
@@ -20,24 +24,28 @@
   font-weight: 300;
 }
 
+.previouslyViewedVersionedStableId {
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.scss:27]: There are things in the name of this class that aren't relevant from the perspective of CSS. The word "versioned", for instance certainly can be omitted. But I am wondering whether the `previouslyViewed` part is significant at all — won't other bookmarks, if we ever have them, be formatted in a similar fashion?

Perhaps `objectId` and `objectType` (or `entityId` and `entityType`) would work better here than `previouslyViewedVersionedStableId` and `previouslyViewedType`?

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.scss [hunk 6]
```diff
@@ -20,24 +24,28 @@
...
+  margin-left: 8px;
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.scss:28]: Notice the problem with the region: an empty span produces an extra space:

image

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx [hunk 7]
```diff
@@ -137,25 +109,23 @@ export const TrackPanelBookmarks = () => {
...
+      {previouslyViewedObjects.length ? (
         <>
-          <div data-test-id="previously viewed" className={styles.sectionTitle}>
-            Previously viewed
-            {previouslyViewedObjects.length > 20 && (
-              <span className={styles.ellipsis}>
+          <PreviouslyViewedLinks />
+          {previouslyViewedObjects.length > 20 && (
+            <div className={styles.more} onClick={onEllipsisClick}>
+              <span>more</span>
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:119]: I tried to see the "more" by switching between over 20 genes, but couldn't. Are you sure you are not removing previously viewed objects from the list if it gets longer than 20 items?

image

Comment (azangru) [reply src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:119]: Also, notice on the screenshot above that you are sometimes repeating stable ids twice (one unversioned, the other versioned). I believe the correct logic is to check if a gene has a symbol, and if it doesn't, then just display a versioned stable id.

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx [hunk 8]
```diff
@@ -137,25 +109,23 @@ export const TrackPanelBookmarks = () => {
...
+              <div className={styles.ellipsis}>
```
[the remark this answers is not in the document]
Comment (azangru) [reply src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:120]: Perhaps the intention was to make the word "more" also clickable?

File: src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts [hunk 9]
```diff
@@ -16,7 +16,7 @@
 
 import { createAction } from 'typesafe-actions';
 import { ThunkAction } from 'redux-thunk';
-import { Action } from 'redux';
+import { Action, ActionCreator } from 'redux';
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts:19]: No need to use the `ActionCreator` type

File: src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts [hunk 10]
```diff
@@ -144,22 +134,29 @@ export const updatePreviouslyViewedObjectsAndSave = (): ThunkAction<
...
+    label: activeEnsObject.label,
+    versioned_stable_id:
+      activeEnsObject.type === 'gene'
+        ? activeEnsObject.versioned_stable_id
+        : null
+  };
+
+  const updatedObject = [newObject, ...savedEntitiesWithoutCurrentEntity].slice(
+    0,
+    20
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts:155]: So how will you ever get past 20 previously viewed items? 😃

File: src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/modal-views/EntityViewerBookmarks.scss [hunk 11]
```diff
@@ -16,6 +16,10 @@
 
 .title{
   font-size: 14px;
-  margin: 5px 0 10px 10px;
+  margin: 5px 0 25px 10px;
   font-weight: $bold;
 }
+
+.versionedStableId {
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/modal-views/EntityViewerBookmarks.scss:23]: Shouldn't matter for CSS that it's versioned

File: src/ensembl/src/content/app/species-selector/state/speciesSelectorEpics.ts [hunk 12]
```diff
@@ -47,14 +47,18 @@ export const fetchSpeciesSearchResultsEpic: Epic<Action, Action, RootState> = (
...
             string
-          >).payload
+          >
+        ).payload ===
+          (
+            action2 as PayloadAction<
+              'species_selector/species_search_request',
+              string
+            >
+          ).payload
```
Comment (azangru) [inline src/ensembl/src/content/app/species-selector/state/speciesSelectorEpics.ts:61]: Is this just prettier? If so, could you please revert this change to avoid confusion?

File: src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts [hunk 13]
```diff
@@ -68,31 +68,28 @@ export const setSearchText = createAction(
   'species_selector/set_search_text'
 )<string>();
```
Comment (azangru) [inline src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts:69]: I'm confused. Why do you need to change anything about species selector actions in order to update the bookmarks behaviour?

File: src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts (hunk 4 above)
Comment (azangru) [reply src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts:143]: This action is not a question 😃

File: src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts (hunk 13 above)
Comment (jyothishnt) [reply src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts:69]: I have no idea how this came in. May be linter is the culprit?

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx (hunk 7 above)
Comment (jyothishnt) [reply src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:119]: Yeah this was one of my doubt to Andrea.

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (jyothishnt) [reply src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:120]: I need to style "more" anyway, so just added the styles for ImageButton there.  Do you think it is better to pass it to ImageButton?

File: src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts (hunk 3 above)
Comment (jyothishnt) [reply src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts:142]: This is a newly created previously viewed object.

File: src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts (hunk 2 above)
Comment (jyothishnt) [reply src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts:153]: Thats true. funny me. This is why we need extra pair of eyes  LOLL

File: src/ensembl/src/content/app/species-selector/state/speciesSelectorEpics.ts (hunk 12 above)
Comment (jyothishnt) [reply src/ensembl/src/content/app/species-selector/state/speciesSelectorEpics.ts:61]: Probably.  Will do it.

File: src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts [hunk 14]
```diff
@@ -143,175 +140,163 @@ export const clearSelectedSearchResult = createAction(
...
+  (genomeId: string): ThunkAction<void, any, null, Action<string>> =>
+  (dispatch, getState: () => RootState) => {
+    const state = getState();
+
+    const currentSpecies = getCommittedSpeciesById(state, genomeId);
+    if (!currentSpecies || currentSpecies.isEnabled) {
+      return;
+    }
+
+    dispatch(toggleSpeciesUseAndSave(genomeId));
```
[the remark this answers is not in the document]
Comment (jyothishnt) [reply src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts:177]: These changes are from a merge commit. Should I worry about this?

[the remark this answers is not in the document]
Comment (azangru) [reply src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts:177]: > These changes are from a merge commit. Should I worry about this?

Well, they are showing up in the diff now 😞 

Could you do `git checkout dev <path to file>` for the files that were not supposed to change in this PR?

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx (hunk 7 above)
Comment (jyothishnt) [reply src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:119]: There is a bug...

File: src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts (hunk 4 above)
Comment (imransl) [reply src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts]: @azangru I didn't check the entire function properly and where it is used. My bad. However, why have we prefixed the function with `ensure` rather than `save`?

Comment (azangru) [reply src/ensembl/src/content/app/species-selector/state/speciesSelectorActions.ts]: Hard to explain 😄  My intention in the naming of this function was: in certain situations (for example when user lands on a page by entering a url in the address bar) a species can be already committed, but we cannot be sure, but we really need it to be. Same with a neighbouring function: a species may be already enabled, but we cannot be sure, but we want to make sure that it is.

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx (hunk 7 above)
Comment (azangru) [reply src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:121]: I can see the "more" link now; but the stable ids are still repeated if a gene does not have a symbol.

image

It might also be worth investigating whether there is anything that can/should be done to the styles of the previously viewed links in the drawer. Currently we crop long lines, which may look weird:

image

I understand that there are several CSS requirements at once that we are trying to balance here (the number of the columns, etc; so there may not be a good CSS solution for the problem that we have).

Finally, notice that there are two vertical scroll bars in the sidebar. This is probably due to how the modal is implemented. Ideally, we should only have a single scrollbar in the sidebar area.

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.scss [hunk 15]
```diff
@@ -13,31 +17,35 @@
...
-  margin-bottom: 15px;
   position: relative;
-  margin-top: 20px;
+  margin-top: 25px;
+  display: inline-flex;
+  cursor: pointer;
+  color: $blue;
+  font-weight: $normal;
 
   .ellipsis {
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.scss:46]: 1) Why are we using an svg ellipsis here rather than a regular unicode ellipsis?
2) If we switch ellipsis to a regular text ellipsis, the `.more` class will have no need for `display: inline-flex`

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.test.tsx [hunk 16]
```diff
@@ -48,15 +49,17 @@ const createRandomPreviouslyViewedObject = (): PreviouslyViewedObject => ({
...
 });
 
 const previouslyViewedObjects = [
   {
     genome_id: genomeId,
     object_id: geneObjectId,
     object_type: 'gene',
-    label: geneId
+    label: geneId,
+    versioned_stable_id: versionedStableId
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.test.tsx:62]: I don't know if there is anything that can be done about it, but I'll just notice here that the `versioned_stable_id` field does not quite fit the information model that we previously had for a `previouslyViewedObject`. Having one field named `label` while the other field named `versioned_stable_id` just feels wrong.

File: src/ensembl/src/content/app/browser/drawer/drawer-views/DrawerBookmarks.tsx [hunk 17]
```diff
@@ -36,10 +37,8 @@ export type DrawerBookmarksProps = {
 };
 
 const DrawerBookmarks = (props: DrawerBookmarksProps) => {
-  const limitedPreviouslyViewedObjects = props.previouslyViewedObjects.slice(
-    0,
-    props.previouslyViewedObjects.length - 20
-  );
+  const limitedPreviouslyViewedObjects =
+    props.previouslyViewedObjects.slice(20);
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/drawer/drawer-views/DrawerBookmarks.tsx:41]: This is the drawer code. Why is there a `slice(20)`? The deleted code did a different job.

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx [hunk 18]
```diff
@@ -82,11 +52,12 @@ export const PreviouslyViewedLinks = () => {
     dispatch(closeTrackPanelModal());
   };
 
+  const limitedPreviouslyViewedObjects = previouslyViewedObjects.slice(0, 20);
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:55]: Is this constant even necessary?

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx [hunk 19]
```diff
@@ -82,11 +52,12 @@ export const PreviouslyViewedLinks = () => {
...
+
   return (
     <div data-test-id="previously viewed links">
-      {[...previouslyViewedObjects]
-        .reverse()
-        .map((previouslyViewedObject, index) => {
+      {[...limitedPreviouslyViewedObjects].map(
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:59]: why the spread?

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx (hunk 18 above)
Comment (jyothishnt) [reply src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:55]: why? You mean using the slice directly on the map? I thought it is good for readability

File: src/ensembl/src/content/app/browser/drawer/drawer-views/DrawerBookmarks.tsx (hunk 17 above)
Comment (jyothishnt) [reply src/ensembl/src/content/app/browser/drawer/drawer-views/DrawerBookmarks.tsx:41]: we always limit to 20 previously viewed object. previous code was working differently

**Comment by azangru** [comment]
Previously viewed objects in the drawer are displayed incorrectly (no space between symbol and stable id):

image

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx [hunk 20]
```diff
@@ -137,25 +104,16 @@ export const TrackPanelBookmarks = () => {
...
-                  onClick={onEllipsisClick}
-                />
-              </span>
-            )}
-          </div>
           <PreviouslyViewedLinks />
+          {previouslyViewedObjects.length > 20 && (
+            <div className={styles.more} onClick={onMoreClick}>
+              more...
+            </div>
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:115]: This whole line, rather than just the word "more..." becomes clickable:






I think you want a span inside of the div; and the styles and click handler should be put on that span:

File: src/ensembl/src/content/app/browser/track-panel/trackPanelState.ts [hunk 21]
```diff
@@ -21,8 +21,8 @@ import pick from 'lodash/pick';
 export type PreviouslyViewedObject = {
   genome_id: string;
   object_id: string;
-  object_type: string;
-  label: string;
+  type: string;
+  label: (string | null)[];
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/trackPanelState.ts:25]: Why did you prefer it over the type `string | string[]` (string or array of strings?)? And what's with the `null[]` type?

File: src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/modal-views/EntityViewerBookmarks.test.tsx [hunk 22]
```diff
@@ -40,11 +40,13 @@ const exampleEntities = [
 
 const previouslyViewedEntities = [
   {
+    versioned_stable_id: 'human-fry.1',
     stable_id: 'human-fry',
     label: 'FRY',
     type: 'gene'
   },
   {
+    versioned_stable_id: 'human-tp53.1',
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/modal-views/EntityViewerBookmarks.test.tsx:49]: What are these two?

File: src/ensembl/src/content/app/browser/drawer/drawer-views/DrawerBookmarks.tsx (hunk 17 above)
Comment (azangru) [reply src/ensembl/src/content/app/browser/drawer/drawer-views/DrawerBookmarks.tsx:41]: Sorry, I forgot the syntax of `Array.slice` for a moment there 😃 I thought this code is limiting previouslyViewedObjects to just 20 items; while in fact it is removing the first 20 items. All good.

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx (hunk 18 above)
Comment (azangru) [reply src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.tsx:50]: I'm slightly bothered by the word `limited` — to me, it makes the name of the variable more confusing than helpful. One other thought is that we could have done the slice back on line 36 (`useSelector(getActiveGenomePreviouslyViewedObjects).slice(0, 20)`). But this is just a stylistic remark; if you prefer the name `limitedPreviouslyViewedObjects`, feel free to ignore this comment.

File: src/ensembl/src/content/app/browser/track-panel/trackPanelState.ts (hunk 21 above)
Comment (jyothishnt) [reply src/ensembl/src/content/app/browser/track-panel/trackPanelState.ts:25]: I think you are right. Label is never going to be null. It will be either symbol + stableId or just stableId.

**Comment by azangru** [comment]
I believe Andrea wants us to rename the "Bookmarks" label in the tooltip to "Previously viewed". I guess this PR would be a better place to do this than the in-app search PR.

image

**Comment by jyothishnt** [comment]
> I believe Andrea wants us to rename the "Bookmarks" label in the tooltip to "Previously viewed". I guess this PR would be a better place to do this than the in-app search PR.
> 
Thanks, good catch. Updated now

File: src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.test.tsx [hunk 23]
```diff
@@ -234,7 +220,7 @@ describe('<TrackPanelBookmarks />', () => {
     );
 
     const ellipsisButton = container.querySelector(
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-modal/modal-views/TrackPanelBookmarks.test.tsx:222]: I guess it's no longer an ellipsis button 😃

**Review by azangru — APPROVED** [review]
Probably better rename the ellipsisButton in the test to moreLink or something.
